BIP375: Add test vectors + validator

macgyver13 opened the pull request:
This PR provides `bip-0375/bip375_test_vectors.json` and a reference `bip-0375/validator/validate_psbt.py` for validating Sending Silent Payments with PSBTs.
PSBTs are validated against v2 requirements with BIP-375 rules (building on BIP-352 silent payment derivation and BIP-374 DLEQ proofs).

Changes since last force-push:
- Add vendored dependencies in `bip-0375/deps`: 
	- `secp256k1lab` (ECC operations)
	- `bitcoin_test` (PSBT parsing adapted from Bitcoin Core test framework)
- Implement 4-stage sequential PSBT validation: structure, ECDH coverage + DLEQ correctness, input eligibility, and output script verification
- Expanded test vector coverage:
	- mixed + ineligible inputs
	- output script validation with same scan key / spend key, same scan / different spend key, different scan key
	- introduce P2TR inputs
- Add test vector version field (starting at 1.1)
- Update BIP-0375 Test Vectors section
- Added a Changelog and Version header
- Updated BIP text about eligible inputs

Feedback welcome @andrewtoth @achow101 @theStack

<details>
<h2>Test Runner Output</h2>

Description: BIP-375 Test Vectors
Version: 1.1

Invalid PSBTs: 22
psbt structure: missing PSBT_OUT_SP_V0_INFO field when PSBT_OUT_SP_V0_LABEL set
psbt structure: incorrect byte length for PSBT_OUT_SP_V0_INFO field
psbt structure: incorrect byte length for PSBT_IN_SP_ECDH_SHARE field
psbt structure: incorrect byte length for PSBT_IN_SP_DLEQ field
psbt structure: PSBT_GLOBAL_TX_MODIFIABLE field is non-zero when PSBT_OUT_SCRIPT set for sp output
psbt structure: missing PSBT_OUT_SCRIPT field when sending to non-sp output
ecdh coverage: P2TR input with NUMS internal key cannot derive sp output
ecdh coverage: only one ineligible P2SH multisig input when PSBT_OUT_SCRIPT set for sp output
ecdh coverage: missing PSBT_IN_SP_ECDH_SHARE field for input 0 when PSBT_OUT_SCRIPT set for sp output
ecdh coverage: missing PSBT_IN_SP_DLEQ field for input when PSBT_IN_SP_ECDH_SHARE set
ecdh coverage: missing PSBT_GLOBAL_SP_DLEQ field when PSBT_GLOBAL_SP_ECDH_SHARE set
ecdh coverage: invalid proof in PSBT_IN_SP_DLEQ field
ecdh coverage: invalid proof in PSBT_GLOBAL_SP_DLEQ field
ecdh coverage: missing PSBT_IN_BIP32_DERIVATION field for input when PSBT_IN_SP_DLEQ set
ecdh coverage: output 1 missing ECDH share for scan key with one input / three sp outputs (different scan keys)
ecdh coverage: input 1 missing ECDH share for output 1 with two inputs / two sp outputs (different scan keys)
ecdh coverage: input 1 missing ECDH share for scan key with two inputs / one sp output
input eligibility: segwit version greater than 1 in transaction inputs with sp output
input eligibility: non-SIGHASH_ALL signature on input with sp output
output scripts: PSBT_OUT_SCRIPT does not match derived sp output
output scripts: two sp outputs (same scan / different spend keys) not sorted lexicographically by spend key
output scripts: k values assigned to wrong output indices with three sp outputs (same scan / spend keys)

Valid PSBTs: 19
can finalize: one P2PKH input single-signer
can finalize: two inputs single-signer using global ECDH share
can finalize: two inputs single-signer using per-input ECDH shares
can finalize: two inputs / two sp outputs with mixed global and per-input ECDH shares
can finalize: one input / one sp output with both global and per-input ECDH shares
can finalize: three sp outputs (different scan keys) with multiple global ECDH shares
can finalize: one P2WPKH input / two mixed outputs - labeled sp output and BIP 32 change
can finalize: one input / two sp outputs - output 0 has no label / output 1 uses label=0 convention for sp change
can finalize: two sp outputs - output 0 uses label=3 / output 1 uses label=1
can finalize: two inputs using per-input ECDH shares - only eligible inputs contribute shares (P2SH excluded)
can finalize: two inputs using global ECDH share - only eligible inputs contribute shares (P2SH excluded)
can finalize: two mixed input types - only eligible inputs contribute ECDH shares (NUMS internal key excluded)
can finalize: three sp outputs (same scan key) - each output has distinct k value
can finalize: three sp outputs (same scan key) / two regular outputs - k values assigned independently of output index
in progress: two P2TR inputs, neither is signed
in progress: one P2TR input / one sp output with no ECDH shares when PSBT_OUT_SCRIPT field is not set
in progress: two inputs / one sp output, input 1 missing ECDH share when PSBT_OUT_SCRIPT field is not set
in progress: one input / two sp outputs, input 0 missing ECDH share for output 0 when PSBT_OUT_SCRIPT field is not set
in progress: large PSBT with nine mixed inputs / six outputs - some inputs signed

Summary: 41 passed, 0 failed
</details>

Test vector generator is available in an [external repo](https://github.com/macgyver13/bip375-test-generator/)

nymius commented on bip-0375/test_vectors.json:
```diff
@@ -0,0 +1,833 @@
+    },
+    {
+      "description": "Mixed segwit versions with silent payments",
+      "psbt": "cHNidP8B+wQCAAAAAQIEAgAAAAEEBAEAAAABBQQBAAAAAQYBAwABDiCuv6p7LnQnsTi6F7UesBJ3TJOXBoHuPhb2Y3hjUklAJAEPBAAAAAABASughgEAAAAAACJSIIYiz2yIzTx5hSOFG+T1WYNgEYY+sDBYObtzXueh9KPpARAE/v///yIGA9NX98BxjyR44/2PjMwnKd3YwMyusfArGBpvRNQ7n42NBAAAAAABAwQBAAAAIh0C0Cn/lt4svPeCvkNZxIYg6pK83WvvAyuVFYuRoWk/tPghAlUWTnkm1Q1SoJ/5kGR6XpXB2xv8aKYW+8LaITkn+Yv/Ih4C0Cn/lt4svPeCvkNZxIYg6pK83WvvAyuVFYuRoWk/tPhAwdZ/ ...
```
I wasn't able to decode this psbt as it is using `bitcoin-cli`, have you?

P.S.: I converted to hex and was able to decode, but it feels as something was off with `bitcoin-cli`.

macgyver13 replied:
First thank you for the review!  
Starting with this comment first, will be able to look at the rest over the next few days.

I had not tested version 2 PSBTs against bitcoin core but there is a version limitation based on my quick scan of psbt.h

- src/psbt.h:L79-80 defines `PSBT_HIGHEST_VERSION = 0`
- src/psbt.h:L1315-1317
```
if (*m_version > PSBT_HIGHEST_VERSION) {
    throw std::ios_base::failure("Unsupported version number");
}
```

nymius replied:
Obviously, if core doesn't support BIP 370, then why `bitcoin-cli` should decode it... well, why not?
I overlooked this, thanks!

nymius reviewed (commented):
cACK 45106322ef3b809e61a13d9b817eb38193fa19fa

- I haven't been able to decode the PSBT's using `bitcoin-cli decodepsbt` in a per test basis. Did you have any issues?

- I would consider copying the approach taken in bitcoin/bips#2084, and use `secp256k1lab` here.

- As each PSBT change is reflected in BIP 174, maybe is worth moving the parsing, fields and PSBT validation logic to BIP 174 directory, and keep the role logic of each protocol on its own directory, like here.

nymius commented:
> Interesting idea. Are you suggesting a storage change for this PR or expand the scope of the PSBT parsing, fields and validation to v0 and other v2 fields and move to other BIPS directories?

I'm suggesting the second one, but as it seems to involve more components, I wouldn't pursue it in this PR.

macgyver13 committed:
BIP-375: Add bitcoin test framework as dependency - deps/bitcoin_test

macgyver13 committed:
BIP-375: add test_runner and validate PSBT structure

Implement psbt structure checks
Add test_runner.py for processing test vectors

macgyver13 committed:
BIP-375: add test vector file

macgyver13 committed:
BIP-375: add BIP375PSBT extension classes

BIP375PSBT (a PSBT subclass that deserializes into BIP375PSBTMap instances)
BIP375PSBTMap (a PSBTMap subclass with BIP-375 field access helpers)

macgyver13 committed:
BIP-375: add ecdh coverage validation

Add deps/dleq.py (Adapted from bip-0374/reference.py)
Extract pubkey from PSBT inputs 
- PSBT_IN_BIP32_DERIVATION
- PSBT_IN_WITNESS_UTXO for P2TR
Add script type helpers
- bip352 input eligibility helpers

macgyver13 committed:
BIP-375: add input eligibility validation

Verify segwit version >1 not used if silent payment outputs present (bip352)
Verify SIGHASH_ALL requirement

macgyver13 committed:
BIP-375: add output scripts validation

Add support for computing bip352 output scripts
Extract ECDH shares and public key from PSBT and aggregate both if necessary
Refactor validate_ecdh_coverage to use collect_input_ecdh_and_pubkey

macgyver13 committed:
BIP-375: update documentation

Update Test Vectors section
Add README.md to explain validation tooling and dependencies

macgyver13 committed:
Merge commit '96000a36c22f6528e834c54f0d115db675198e57' as 'bip-0375/deps/secp256k1lab'

macgyver13 committed:
Squashed 'bip-0375/deps/secp256k1lab/' content from commit 44dc4bd

git-subtree-dir: bip-0375/deps/secp256k1lab
git-subtree-split: 44dc4bd893b8f03e621585e3bf255253e0e0fbfb

macgyver13 committed:
BIP-375: address review feedback

correctly label witness_utxo vs non_witness_utxo key in supplementary inputs

Summary of test vector changes:
removed test: 
- psbt structure: empty PSBT_OUT_SCRIPT field when sending to non-sp output
modified test:
- ecdh coverage: only one ineligible P2SH multisig input when PSBT_OUT_SCRIPT set for sp output
- can finalize: one P2PKH input single-signer
- can finalize: two inputs using per-input ECDH shares - only eligible inputs contribute shares (P2SH excluded)
added test: 
- can finalize: two inputs using global ECDH share - only eligible inputs contribute shares (P2SH excluded)

macgyver13 committed:
BIP-375: clarify eligible input restriction in Signer text

macgyver13 commented on bip-0375.mediawiki:
```diff
@@ -177,7 +177,7 @@ All rules must be followed from PSBTv2 for this role. If there are any outputs w
 
 For each output with PSBT_OUT_SP_V0_INFO set, the Signer should:
-* Compute and set an ECDH share and DLEQ proof for each input it has the private key for, or set a global ECDH share and DLEQ proof if it has private keys for all eligible inputs.
+* Compute and set an ECDH share and DLEQ proof for each eligible input it has the private key for, or set a global ECDH share and DLEQ proof if it has private keys for all eligible inputs.
```
After reflecting on the previous version of the Signer text change I realized I had inserted a 'must' bullet into a 'should' section.  I think the addition of eligible as a 'should' resolves the ambiguity and I don't see this as a backward incompatible change.

I chose to isolate this change in a single commit for easy review and linkage if necessary.

@andrewtoth - If you think a Mailing List Update is warranted let me know.

andrewtoth replied:
Can we have a test vector that validates a PSBT that has shares on both eligible and ineligible inputs? As per the spec now, this test should pass by ignoring the shares on ineligible inputs.

Producers should not add these shares, but the consumer should also not fail if they are present.

macgyver13 committed:
BIP-375: skip ineligible inputs when combining ecdh shares

add fake ecdh share and dleq proof to P2SH input for valid test: two inputs using per-input ECDH shares - only eligible inputs contribute shares (P2SH excluded)

remove unused return string from is_input_eligible

macgyver13 committed:
BIP-375: clarify eligible inputs restriction in Computing the Output Scripts text

macgyver13 committed:
BIP-375: fix label byte order used by labelhash

Test vectors with labels now use big-endian byte order instead of little-endian, matching BIP-352 specification

Summary of test vector changes:
- psbt structure: missing PSBT_OUT_SP_V0_INFO field when PSBT_OUT_SP_V0_LABEL set
- can finalize: one P2WPKH input / two mixed outputs - labeled sp output and BIP 32 change
- can finalize: two sp outputs - output 0 uses label=3 / output 1 uses label=1

andrewtoth commented:
ACK b217897a628e3d5db369497d2697f76e5bab7f4d

murchandamus commented:
Since this PR has been signed off by one author, I will slate this PR for merging. @theStack, @achow101, please take a look in the next few days if you want to review before merging.

macgyver13 commented:
> Since this PR has been signed off by one author, I will slate this PR for merging. @theStack, @achow101, please take a look in the next few days if you want to review before merging.

Thanks! I have held off on rebasing during review to keep commits stable. Could you give me a heads up before merging so I can do a final cleanup pass?

murchandamus commented:
@macgyver13: I usually do a merge commit when the commits are well-structured like here, and I squash the commits when it’s a topsy-turvy collection of omnibus commits and typo fix commits. I was gonna merge this PR before the end of the week if nothing else were to happen, but I’m happy to wait for your cue that it’s in the shape you want it.

jonatack commented on bip-0375.mediawiki:
```diff
@@ -249,7 +249,156 @@ Silent payment capable PSBTs are backwards compatible with PSBTv2 once all outpu
 ==Test Vectors==
 
-Todo
+A [[bip-0375/bip375_test_vectors.json|collection of test vectors in JSON format]] is provided. Each test vector contains a base64-encoded PSBT string, which alone can be used to verify sending Silent Payments with PSBTs.
```
Consider adding a changelog to the BIP (as done in this pull for [secp256k1lab](https://github.com/bitcoin/bips/pull/2046/changes#diff-661f1bb5001eb83c34e4d1e306d28d4e96d5d4f65648df1c3ec78db19fe31c52R1)) with a minor release update for adding the test vectors.

murchandamus replied:
Slight amendment: BIP3 [recommends](https://github.com/bitcoin/bips/blob/master/bip-0003.md#changelog-section-and-version-header) bumping PATCH for a change such as adding test vectors. The Changelog section is also optional before a BIP is advanced to Complete. Should you add a Changelog, please also add a Version header.

<img width="1026" height="240" alt="image" src="https://github.com/user-attachments/assets/627840fd-833b-4157-95f9-9cc1e1b0af30" />

macgyver13 replied:
Changelog Section added - let me know if I need to make more tweaks

andrewtoth commented:
cc @craigraw I believe you implemented BIP375 in sparrow wallet.

macgyver13 committed:
BIP-375: add changelog section

Add Changelog section
Begin with version 0.1.0 as this BIP is Draft phase

macgyver13 commented:
> @macgyver13: I usually do a merge commit when the commits are well-structured like here, and I squash the commits when it’s a topsy-turvy collection of omnibus commits and typo fix commits. I was gonna merge this PR before the end of the week if nothing else were to happen, but I’m happy to wait for your cue that it’s in the shape you want it.

@murchandamus Thank you for the perspective.  I am comfortable leaving the commits as-is based on your feedback. I consider this PR ready for merge when you see fit.

jonatack commented on bip-0375.mediawiki:
```diff
@@ -400,6 +400,18 @@ Use the provided [[bip-0375/test_runner.py|test runner]] to validate each test v
+To help implementers understand updates to this document, we attach a version number that resembles ''semantic versioning'' (<code>MAJOR.MINOR.PATCH</code>).
+The <code>MAJOR</code> version is incremented if changes to the BIP are introduced that are incompatible with prior versions.
+The <code>MINOR</code> version is incremented whenever the inputs or the output of an algorithm changes in a backward-compatible way or new backward-compatible functionality is added.
+The <code>PATCH</code> version is incremented for other changes that are noteworthy (bug fixes, test vectors, important clarifications, etc.).
```
No need for these lines of boilerplate. (Can squash the removal commit into the last commit [`d447a19`](https://github.com/bitcoin/bips/pull/2046/commits/d447a197fb99175e1d5b8745ef2175d13c1f6d00) that adds them.)

jonatack reviewed (commented):
> so I can do a final cleanup pass

@macgyver13 I agree the squash and merge commits could be cleaned up, among any others you had in mind.

murchandamus commented on bip-0375.mediawiki:
Please also add the Version header corresponding to the Changelog in the Preamble:

```diff
 Discussion: https://groups.google.com/g/bitcoindev/c/5G5wzqUXyk4
+Version: 0.1.1
 Requires: 352, 370, 374
```

murchandamus reviewed (commented):
LGTM, one amendment requested

murchandamus commented:
[Changes since @andrewtoth’s ACK](https://github.com/bitcoin/bips/compare/b217897a628e3d5db369497d2697f76e5bab7f4d..8c0a9bfa57c515205fa07851128d758ef72e433b) are only the addition of the Changelog and Version header.

andrewtoth commented:
ACK 8c0a9bfa57c515205fa07851128d758ef72e433b

jonatack commented on bip-0375.mediawiki:
```diff
@@ -249,7 +250,163 @@ Silent payment capable PSBTs are backwards compatible with PSBTv2 once all outpu
+== Changelog ==
+
+* '''0.1.1''' (2026-04-16):
+** Add test vectors and reference for validating Sending Silent Payments with PSBTs.
```
There are also updates to the BIP about eligible inputs / restrictions. Should that be mentioned?

jonatack replied:
Nit (not a blocker for this already substantial and BIP author approved improvement):

Referencing the above comment as well as https://github.com/bitcoin/bips/pull/2046/changes#r3089267494, it seems to me that those updates, along with the new addition of the Valid PSBTs section, might suggest incrementing a minor, rather than patch, version (e.g., to 0.2.0), if the specification is considered to be modified/extended (in a backward-compatible way).

(Both of these comments can be handled in a follow-up PR.)

murchandamus replied:
I read the addition of "eligible" as a simple clarification, because only some inputs are relevant to Silent Payments.

jonatack commented:
Appended the following to the PR description (@macgyver13 feel free to improve/tweak):

```
- Added a Changelog and Version header
- Updated BIP text about eligible inputs
```

nothingmuch commented:
> I believe if present, the PSBT_OUT_SCRIPT should have a valid script. So, empty is invalid.

an empty scriptPubKey is non-standard but in consensus is valid and spendable by a scriptSig that pushes a true value
